Enable finegrained diffs, update() args include desired state, latest state

kumargauravsharma opened the pull request:
Issue #, if available: #298 

**Description of changes:**
Changed the update() method args to include 'latest' state as well. 
Now both desired state and latest state is available inside update()
method to compute finegrained diffs to make service api calls.

Also, it addresses comment received on #245 to simplify 
the mechanism to configure custom implementation for operation 
on a custom resource. Following is an example to 
configure `CustomModifyReplicationGroup` as custom_implementation operation 
for `ModifyReplicationGroup` operation in elasticache `generator.yaml`
config file:
```
operations:
  ModifyReplicationGroup:
    custom_implementation: CustomModifyReplicationGroup
```

**Due to open issue #296 not all services' controllers are auto generated in this
revision of the PR. Thus, `make test` fails for other services as of now.**
> This PR includes service controllers generated for `elasticache` and
> manually updated for `petstore`, `bookstore` services to help this review.
> Once issue #296 is fixed, remaining service controller will be generated
> and this PR will include corresponding updates.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

nmvk reviewed (commented):
I liked this approach as it would be easy to add status checks to with this approach.

jaypipes commented on services/elasticache/pkg/resource/replication_group/custom_update_api.go:
```diff
@@ -87,30 +107,32 @@ func (rm *resourceManager) UpdateReplicaCount(
 func (rm *resourceManager) newUpdateShardConfigurationRequestPayload(
-	r *resource,
+	rd *resource, // desired
+	rl *resource, // latest
```
actually, I was originally confused why you were passing the `resourceDescriptor` but then realized that you were just calling the "desired" resource parameter `rd`. Recommend calling these parameters `desired` and `latest` :) We use "rd" to indicate the `resourceDescriptor` struct

kumargauravsharma replied:
Renamed variables.

jaypipes commented on services/petstore/pkg/resource/pet/manager.go:
```diff
@@ -150,15 +150,16 @@ func (rm *resourceManager) newCreateRequestPayload(
 	ctx context.Context,
-	res acktypes.AWSResource,
+	resDesired acktypes.AWSResource, // desired
+	resLatest acktypes.AWSResource, // latest
```
FYI, we're getting rid of the petstore and bookstore services here:

https://github.com/aws/aws-controllers-k8s/pull/305

So you won't need this after you rebase.

kumargauravsharma replied:
Rebased with latest from main.

jaypipes reviewed (changes requested):
Small renaming requested inline. Also, you can delete the petstore/bookstore modifications since we're killing those off here: https://github.com/aws/aws-controllers-k8s/pull/305

kumargauravsharma commented:
> Small renaming requested inline. Also, you can delete the petstore/bookstore modifications since we're killing those off here: #305
Updated code and rebased with latest from main.

jaypipes commented on services/elasticache/generator.yaml:
```diff
@@ -3,16 +3,9 @@ resources:
-          - Spec.ReplicasPerNodeGroup
 operations:
   ModifyReplicationGroup:
+    custom_implementation: CustomModifyReplicationGroup
```
:+1:

jaypipes reviewed (approved):
\o/ :)

jaypipes commented:
I'm confused why this particular patch is failing make test in the way it is (seems a mockery error).

kumargauravsharma commented:
Service controllers need to be auto generated due to change in method signatures for
`manager::update` and `sdk::sdkUpdate` that this PR brings.

Due to open issue #296, not all services' controllers are auto generated in this PR. 
Thus, make test fails for other services.

kumargauravsharma commented:
rebased with main.

jaypipes reviewed (approved):
alright, let's do this. ++

jaypipes commented:
@kumargauravsharma just a quick note, I'm going to squash and merge this, but in the future, please do `git commit --amend && git push origin <BRANCH> --force` if you make changes to a commit based on a code review instead of pushing more commits with commit messages like "fixed after rebase" etc :)
